freqtrade/exchange/kraken.py

```diff
@@ -95,32 +108,82 @@ def stoploss(self, pair: str, amount: float, stop_price: float, order_types: Dic
         except ccxt.InvalidOrder as e:
             raise InvalidOrderException(
-                f'Could not create {ordertype} sell order on market {pair}. '
+                f'Could not create {ordertype} {side} order on market {pair}. '
                 f'Tried to create stoploss with amount {amount} at stoploss {stop_price}. '
                 f'Message: {e}') from e
         except ccxt.DDoSProtection as e:
             raise DDosProtection(e) from e
         except (ccxt.NetworkError, ccxt.ExchangeError) as e:
             raise TemporaryError(
-                f'Could not place sell order due to {e.__class__.__name__}. Message: {e}') from e
+                f'Could not place {side} order due to {e.__class__.__name__}. Message: {e}') from e
         except ccxt.BaseError as e:
             raise OperationalException(e) from e
+
+    def fill_leverage_brackets(self):
+        """
+            Assigns property _leverage_brackets to a dictionary of information about the leverage
+            allowed on each pair
+        """
+        leverages = {}
+
+        for pair, market in self.markets.items():
+            leverages[pair] = [1]
+            info = market['info']
+            leverage_buy = info.get('leverage_buy', [])
+            leverage_sell = info.get('leverage_sell', [])
+            if len(leverage_buy) > 0 or len(leverage_sell) > 0:
+                if leverage_buy != leverage_sell:
+                    logger.warning(
+                        f"The buy({leverage_buy}) and sell({leverage_sell}) leverage are not equal"
+                        "for {pair}. Please notify freqtrade because this has never happened before"
+                    )
+                    if max(leverage_buy) <= max(leverage_sell):
+                        leverages[pair] += [int(lev) for lev in leverage_buy]
+                    else:
+                        leverages[pair] += [int(lev) for lev in leverage_sell]
+                else:
+                    leverages[pair] += [int(lev) for lev in leverage_buy]
+        self._leverage_brackets = leverages
+
+    def get_max_leverage(self, pair: Optional[str], nominal_value: Optional[float]) -> float:
+        """
+            Returns the maximum leverage that a pair can be traded at
+            :param pair: The base/quote currency pair being traded
+            :nominal_value: Here for super class, not needed on Kraken
+        """
+        return float(max(self._leverage_brackets[pair]))
+
+    def _set_leverage(
+        self,
+        leverage: float,
+        pair: Optional[str] = None,
+        trading_mode: Optional[TradingMode] = None
+    ):
+        """
+            Kraken set's the leverage as an option in the order object, so we need to
+            add it to params
+        """
+        if leverage > 1.0:
+            self._params['leverage'] = leverage
+        else:
+            if 'leverage' in self._params:
+                del self._params['leverage']
```

Comment:
This is wrong and pretty dangerous (to produce random failures).
It's also causing the random failures (it's a mutable class-object again - so not resetting this will cause test-leakage).

Obviously you'll need to then overwrite that method for kraken, then ...

setting this at instance level (outside of create_order) will fail (badly) in case there is a force-sell (or forcebuy) at the same time than regular order-placement.
As such,  you cannot rely on the fact that context remains within the method, as the stuff from RPC comes in from a different thread.

So far, this isn't a problem because there's no "global" setting that's being set which affects order placement (and for exits there's a lock in place) - but for buys it's for sure an issue (and can cause leverage settings to be mixed up between the "regular" and forcebuy orders).


In my opinion (and understanding), the parameter "leverage" should be added to `create_order()` (i think i mentioned this before - but either you didn't understand that, or tried to ignore that).

The you can have a lock within the exchange that's protecting us from these cases by waiting for the "other" order to be finalized.

it'll also be easier to manage overall, as the whole "set_leverage" stuff will not cause 2 calls to the exchange class, but will be handled implicitly by the exchange.

Reply:
> In my opinion (and understanding), the parameter "leverage" should be added to create_order() (i think i mentioned this before - but either you didn't understand that, or tried to ignore that).

It is a parameter

Reply:
> Also, whenever i've looked in here i've seen new commits, so i assumed it's not complete / done (i assumed so the weekend, but then new commits started to appear, adding new stuff ...).

Yeah I kept thinking of changes that needed to be made, i think I got them all